### PR TITLE
First pass at chat

### PR DESCRIPTION
**Pull request by chrisnojima**
This is a first pass at chat, just trying to get it functionally working end to end
- [x] Simple text messages
- [x] Simple inbox
- [x] Marcos' infinite scroll list for conversations
- [x] Load more when scrolling to the top
- [x] Listen to gregors to append to the bottom

A lot of know issues. Messages can be posted out of order. Inboxes don't refresh etc

As before I split the commit into several sections
### 1: https://github.com/keybase/client/pull/4743/commits/05df6d105512d14d54b74e3a750c8e4e082eb0d4 Misc stuff
- [x] Updated dependencies
- [x] Support IdentifyUI cancel
- [x] Plumb through some SmallInput params
- [x] Engine allows cancelSession
- [x] Engine log shows method names so you can find sent/response from the server easier
- [x] Global crash handler will rethrow in **DEV**
- [x] Chat tabs / feature flags
### 2: https://github.com/keybase/client/pull/4743/commits/8b7d22aad16516c999e8b6811d8b7b70a6d8eb85 Code gen
- [x] Cleanup of bin/flow
- [x] Generated code now includes full namespace in calls as we now have keybase.1 and chat.1 methods
### 3: https://github.com/keybase/client/pull/4743/commits/657dc46a8ede6f09f57507e2ebcb75935fef9e34 Chat redux
- [x] Chat reducer
- [x] Chat actions
- [x] Chat view skeleton
### 4: https://github.com/keybase/client/pull/4743/commits/7b0d41737a34ed5998cbce65f81114143dba8c73 A conversation
- [x] The conversation view: Header, List, Input
- [x] Loading more type, plain text type
### 5: https://github.com/keybase/client/pull/4743/commits/81748674746a6b7fea5f66f2af6c4ce84c69eb40 The inbox
- [x] The inbox view (conversations list)

### CODE REVIEW

**Comment by chrisnojima** [comment]
@keybase/react-hackers 👀

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump

File: shared/actions/chat.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+      // $FlowIssue doens't understand try
+      snippet = _makeSnippet(recentMessage.valid.messageBody.text.body, 100)
+    } catch (_) { }
+
+    return new InboxStateRecord({
+      info: convo.info,
+      conversationIDKey: conversationIDToKey(convo.info.id),
+      participants: List(convo.info.writerNames || []) || List(), // TODO in recent order... somehow
+      muted: false,
+      time: 'Time',
```
Comment (MarcoPolo) [inline shared/actions/chat.js]: This is a placeholder right? maybe add a todo

File: shared/chat/conversation/list.desktop.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+      isScrolling: true,
+      scrollTop,
+    })
+
+    // This is debounced so it resets the call
+    this._onScrollSettled()
+  }, 100)
+
+  _rowRenderer = ({index, key, style, isScrolling}: {index: number, key: string, style: Object, isScrolling: boolean}) => {
+    if (!index) {
```
Comment (MarcoPolo) [inline shared/chat/conversation/list.desktop.js]: 🐑  I think this is more readable as `index === 0`

File: shared/actions/chat.js [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+      const messageUnboxed: MessageUnboxed = incomingMessage.message
+      const yourNameSelector = (state: TypedState) => state.config.username
+      const yourName = yield select(yourNameSelector)
+      const message = _threadToStorable(messageUnboxed, 0, yourName)
+
+      yield put({
+        type: Constants.appendMessages,
+        payload: {
+          conversationIDKey: conversationIDToKey(incomingMessage.convID),
+          messages: [message],
```
Comment (MarcoPolo) [inline shared/actions/chat.js]: I think in general we don't want to log message contents. We really don't want to see messages in debug logs.

File: shared/chat/conversation/messages/loading-more.desktop.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// @flow
+import React from 'react'
+import {Box, Text} from '../../../common-adapters'
+import {globalStyles} from '../../../styles'
+
+import type {Props} from './loading-more'
+
+const MessageLoadingMore = ({style, loading}: Props) => (
```
Comment (MarcoPolo) [inline shared/chat/conversation/messages/loading-more.desktop.js]: I think you should change the name of `loading` to `hasMoreItems`, I was a bit confused

File: shared/chat/conversations-list/index.js.flow [hunk 5]
```diff
@@ -0,0 +1,12 @@
+// @flow
+import {Component} from 'react'
+import {List} from 'immutable'
+
+import type {InboxStateRecord, ConversationIDKey} from '../../constants/chat'
+
+export type Props = {
+  inbox: List<InboxStateRecord>,
```
Comment (MarcoPolo) [inline shared/chat/conversations-list/index.js.flow]: Should be `List<InboxState>`

File: shared/actions/tracker.js [hunk 6]
```diff
@@ -607,6 +607,10 @@ function _serverCallMap (dispatch: Dispatch, getState: Function, isGetProfile: b
         expiringLocal: false,
       })
     },
+    'keybase.1.identifyUi.cancel': ({sessionID}, response) => {
+      response.result()
+      engine().cancelSession(sessionID)
+    },
```
Comment (chrisnojima) [inline shared/actions/tracker.js]: let's ignore this for now, shouldn't affect anything

File: shared/constants/chat.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+import {Buffer} from 'buffer'
+
+import type {ConversationID as RPCConversationID, ChatActivity, ConversationInfoLocal} from './types/flow-types-chat'
+import type {NoErrorTypedAction} from './types/flux'
+
+export type MessageType = 'Text'
+export type FollowState = 'You' | 'Following' | 'Broken' | 'NotFollowing'
+
+export type Message = {
+  type: 'Text',
```
Comment (chrisnojima) [inline shared/constants/chat.js]: ? this is me describing the various types.  so MessageType will actually to be a list so i think this should stay as is

File: shared/constants/chat.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+export const loadMoreMessages = 'chat:loadMoreMessages'
+export const loadingMessages = 'chat:loadingMessages'
+export const prependMessages = 'chat:prependMessages'
+export const setupNewChatHandler = 'chat:setupNewChatHandler'
+export const incomingMessage = 'chat:incomingMessage'
+export const postMessage = 'chat:postMessage'
+
+export type AppendMessages = NoErrorTypedAction<'chat:appendMessages', {conversationIDKey: ConversationIDKey, messages: Array<Message>}>
+export type LoadInbox = NoErrorTypedAction<'chat:loadInbox', void>
+export type LoadedInbox = NoErrorTypedAction<'chat:loadedInbox', {inbox: List<InboxStateRecord>}>
```
Comment (chrisnojima) [inline shared/constants/chat.js]: yup

File: shared/chat/container.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
+// @flow
+import React, {Component} from 'react'
+import {connect} from 'react-redux'
+import Chat from './index'
+
+import {loadInbox} from '../actions/chat'
+
+class ChatContainer extends Component {
+  componentWillMount() {
+    this.props.loadInbox()
```
Comment (chrisnojima) [inline shared/chat/container.js]: lets's leave this for now. currently the inbox doens't get updates correctly so let's wait until thats actually working

File: shared/actions/chat.js [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+  // $FlowIssue flow doesn't understand spread + strings
+  return [...(message.substring(0, max * 4).replace(/\s+/g, ' '))].slice(0, max).join('')
+}
+
+function _inboxToConversations (inbox: GetInboxAndUnboxLocalRes): List<InboxState> {
+  return List((inbox.conversations || []).map(convo => {
+    const recentMessage: ?MessageUnboxed = (convo.maxMessages || []).find(message => (
+      message.state === LocalMessageUnboxedState.valid &&
+      message.valid &&
+      message.valid.messageBody.messageType === CommonMessageType.text
```
Comment (chrisnojima) [inline shared/actions/chat.js]: i just tried to do this but that means i need to map the entire thing or do more logic to find it (keeping this logic) then just converting it to just pull out the string. i think this is very specific to how the inbox works and likely won't occur elsewhere)

File: shared/actions/chat.js [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+    let snippet
+    try {
+      // $FlowIssue doens't understand try
+      snippet = _makeSnippet(recentMessage.valid.messageBody.text.body, 100)
+    } catch (_) { }
+
+    return new InboxStateRecord({
+      info: convo.info,
+      conversationIDKey: conversationIDToKey(convo.info.id),
+      participants: List(convo.info.writerNames || []) || List(), // TODO in recent order... somehow
```
Comment (chrisnojima) [inline shared/actions/chat.js]: shouldn't need it tnx

File: shared/actions/chat.js (hunk 1 above)
Comment (chrisnojima) [inline shared/actions/chat.js]: yup

File: shared/actions/chat.js [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+    },
+  })
+}
+
+function * _incomingMessage (action: IncomingMessage): SagaGenerator<any, any> {
+  if (action.payload.activity.ActivityType === NotifyChatChatActivityType.incomingMessage) {
+    const incomingMessage: ?IncomingMessageRPCType = action.payload.activity.IncomingMessage
+    if (incomingMessage) {
+      const messageUnboxed: MessageUnboxed = incomingMessage.message
+      const yourNameSelector = (state: TypedState) => state.config.username
```
Comment (chrisnojima) [inline shared/actions/chat.js]: k

File: shared/actions/chat.js (hunk 3 above)
Comment (chrisnojima) [inline shared/actions/chat.js]: sounds good. adding hiddenstring to Message and actions

File: shared/actions/chat.js [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+  const oldConversationState = yield select(conversationStateSelector)
+
+  let next
+  if (oldConversationState) {
+    if (oldConversationState.get('isLoading')) {
+      __DEV__ && console.log('Bailing on chat load more due to isloading already')
+      return
+    }
+
+    if (!oldConversationState.get('moreToLoad')) {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: This is getting older messages. We can't be confused about this state. Newer messages come through gregor, not this.

File: shared/actions/chat.js [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+    conversationID,
+    query: {markAsRead: true},
+    pagination: {
+      next,
+      num: Constants.maxMessagesToLoadAtATime,
+    },
+  }})
+
+  const yourNameSelector = (state: TypedState) => state.config.username
+  const yourName = yield select(yourNameSelector)
```
Comment (chrisnojima) [inline shared/actions/chat.js]: done

File: shared/actions/chat.js [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+  if (thread && thread.thread && thread.thread.pagination) {
+    return thread.thread.pagination
+  }
+  return {
+    last: undefined,
+    next: undefined,
+  }
+}
+
+function _threadToStorable (message: MessageUnboxed, idx: number, yourName): Message {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: yup, now called _unboxedToMessage

File: shared/chat/conversation/messages/index.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
+// @flow
+import MessageText from './text'
+import React from 'react'
+import {Box} from '../../../common-adapters'
+
+import type {Message} from '../../../constants/chat'
+
+const factory = (message: Message, index: number, key: string, style: Object, isScrolling: boolean) => {
```
Comment (chrisnojima) [inline shared/chat/conversation/messages/index.js]: well this is just a function its not a component

File: shared/chat/conversation/messages/loading-more.desktop.js (hunk 4 above)
Comment (chrisnojima) [inline shared/chat/conversation/messages/loading-more.desktop.js]: k

File: shared/chat/conversation/messages/index.js (hunk 16 above)
Comment (MarcoPolo) [inline shared/chat/conversation/messages/index.js]: ah okay

**Pull request merged**